crypto/stack/stack.c

```diff
@@ -107,43 +147,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
 
     if ((ret = OPENSSL_zalloc(sizeof(*ret))) == NULL)
         goto err;
-    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * MIN_NODES)) == NULL)
+    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * min_nodes)) == NULL)
         goto err;
     ret->comp = c;
-    ret->num_alloc = MIN_NODES;
+    ret->num_alloc = min_nodes;
     return (ret);
 
  err:
     OPENSSL_free(ret);
     return (NULL);
 }
 
-int OPENSSL_sk_insert(OPENSSL_STACK *st, const void *data, int loc)
+int OPENSSL_sk_reserve(OPENSSL_STACK *st, int n)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
+    if (st == NULL || st->num < 0)
         return 0;
-    }
 
-    if (st->num_alloc <= (size_t)(st->num + 1)) {
-        size_t doub_num_alloc = st->num_alloc * 2;
-        const char **tmpdata;
+    if (n <= 0)
+        return 1;
 
-        /* Overflow checks */
-        if (doub_num_alloc < st->num_alloc)
-            return 0;
+    /* Sanity checks */
+    if (n > INT_MAX - st->num
+        || (size_t)n > SIZE_MAX / sizeof(char *) - st->num)
+        return 0;
+
+    if ((size_t)st->num + n >= st->num_alloc) {
```

Comment:
Even here *if* one make assumption that size_t is narrower than int, then you'd be better of with implicit casts.

Reply:
Well, I bet warning system will complain about signed/unsigned comparison, so implicit cast won't do, right? But in such case it would be formally correct to cast left part to unsigned int and let implicit casts play out between unsigned int and size_t.

Reply:
Again, *if* we make such assumption. Though one still argue that unsigned int cast of the left part is more readable.

Reply:
It was a build warning that prompted me to add the cast.
unsigned int seems reasonable and it is guaranteed to work.